core/sequencer/tracker/tracker.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/google/trillian/util/election2"
+)
+
+const resourceLabel = "resource"
+
+var (
+	once     sync.Once
+	isMaster monitoring.Gauge
+)
+
+func createMetrics(mf monitoring.MetricFactory) {
+	isMaster = mf.NewGauge(
+		"is_master",
+		"Set to 1 for resources for which this instance is currently master",
+		resourceLabel)
+}
+
+// Tracker tracks mastership of a collection of resources.
+type Tracker struct {
+	factory     election2.Factory
+	maxHold     time.Duration
+	master      map[string]election2.Election
+	masterMu    sync.RWMutex
+	watching    map[string]bool
+	watchingMu  sync.RWMutex
+	newResource chan string
+}
+
+// New returns a new mastership tracker
+func New(factory election2.Factory, maxHold time.Duration, metricsFactory monitoring.MetricFactory) *Tracker {
+	once.Do(func() { createMetrics(metricsFactory) })
+	return &Tracker{
+		factory:     factory,
+		maxHold:     maxHold,
+		master:      make(map[string]election2.Election),
+		watching:    make(map[string]bool),
+		newResource: make(chan string),
+	}
+}
+
+// AddResource makes the mastership tracker aware of new resources.
+// The same resource may be added an unlimited number of times.
+func (mt *Tracker) AddResource(res string) {
+	mt.newResource <- res
+}
+
+// Run starts new watchers for new resources.
+func (mt *Tracker) Run(ctx context.Context) {
+	for {
+		select {
+		case res := <-mt.newResource:
+			if !mt.isWatching(res) {
+				go mt.watchResource(ctx, res, mt.maxHold)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// watchResource is a blocking method that runs elections for res and updates mt.masterfor.
+func (mt *Tracker) watchResource(ctx context.Context, res string, resign time.Duration) {
+	e, err := mt.factory.NewElection(ctx, res)
+	if err != nil {
+		glog.Errorf("failed to create election for %v: %v", res, err)
+		return
+	}
+	mt.setWatching(res)
+	defer func(ctx context.Context) {
+		if err := e.Close(ctx); err != nil {
+			glog.Warningf("election.Close(%v): %v", res, err)
+		}
+		mt.setNotWatching(res)
+	}(ctx)
+
+	for {
+		if err := func() error {
+			mt.setNotMaster(res)
+			if err := e.Await(ctx); err != nil {
+				return err
+			}
+			glog.Infof("Obtained mastership for %v", res)
+
+			mt.setMaster(res, e)
+			defer mt.setNotMaster(res)
+
+			mastershipCtx, err := e.WithMastership(ctx)
+			if err != nil {
+				return err
+			}
+
+			select {
+			case <-time.After(resign):
+				glog.Infof("Resigning from %v after %v", res, resign)
+				if err := e.Resign(ctx); err != nil {
+					glog.Errorf("Resign(%v): %v", res, err)
+				}
+			case <-mastershipCtx.Done():
+				glog.Warningf("No longer master for %v", res)
+				// If the master ctx is canceled, exit for loop.
+				if err := ctx.Err(); err != nil {
+					return err
+				}
+				if err := e.Resign(ctx); err != nil {
+					glog.Errorf("Resign(%v): %v", res, err)
+				}
+			}
+			return nil
+		}(); err != nil {
+			glog.Errorf("watchResource(%v): %v", res, err)
```

Comment:
nit: Maybe make `watchResource` method return `error`, and log it on the caller's side in line 76? Looks like it will allow you to `return func() error {...} ()` directly, and also delete line 88.